### PR TITLE
feat: mac credential storage

### PR DESCRIPTION
**Pull request by brianointel**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-examples/blob/master/.github/CONTRIBUTING.md.

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

## New Imports
Are there any new imports or modules? If so, what are they used for and why?

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?

## Other information

### CODE REVIEW

File: internal/driver/checkstatuses.go [hunk 1]
```diff
@@ -25,8 +25,24 @@ func (d *Driver) checkStatuses() {
 
 		status := d.testConnectionMethods(device)
 
-		if err := d.updateDeviceStatus(device.Name, status); err != nil {
+		if statusChanged, err := d.updateDeviceStatus(device.Name, status); err != nil {
 			d.lc.Warnf("Could not update device status for device %s: %s", device.Name, err.Error())
+		} else if statusChanged && status == UpWithAuth {
+			d.lc.Infof("Device %s is now %s, refreshing the device information.", device.Name, UpWithAuth)
+			go func(device models.Device) {
+				// todo: should we update the device?
```
Comment (lenny-goodell) [inline internal/driver/checkstatuses.go]: Yes, looks like the Device data changed,  so Core Metadata needs that updated data.
```
device.Protocols[OnvifProtocol][DeviceStatus] = status
```

File: internal/driver/credentials.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+	return credentials, err
+}
+
+func (d *Driver) getCredentialsFromMac(mac string) (Credentials, errors.EdgeX) {
+	if mac == "" {
+		credential, edgexErr := tryGetCredentials(d.config.AppCustom.DefaultSecretPath)
+		if edgexErr != nil {
+			d.lc.Error("failed to get credentials from default secret path", "err", edgexErr)
+			return Credentials{}, errors.NewCommonEdgeX(errors.KindServerError, fmt.Sprintf("failed to get credentials from mac address %s", mac), edgexErr)
```
Comment (lenny-goodell) [inline internal/driver/credentials.go]:
```suggestion
			return Credentials{}, errors.NewCommonEdgeX(errors.KindServerError, "failed to get default credentials for empty mac address), edgexErr)
```

File: internal/driver/credentials.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+		}
+
+		d.lc.Debug("Using default credentials from default secret path")
```
Comment (lenny-goodell) [inline internal/driver/credentials.go]:
```suggestion
		d.lc.Debug("Using default credentials from default secret path for empty mac address")
```

File: internal/driver/driver.go [hunk 4]
```diff
@@ -652,6 +616,23 @@ func addressAndPort(xaddr string) (string, string) {
 	}
 }
 
+// todo: this should be integrated better with getDeviceInformation to avoid creating another temporary client
```
Comment (lenny-goodell) [inline internal/driver/driver.go:619]: Yes! Why was this not done now?

File: internal/driver/macmapper_test.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	for _, test := range tests {
+		test := test
+		t.Run(test.mac, func(t *testing.T) {
+			sanitized, err := SanitizeMACAddress(test.mac)
+			assert.Equal(t, test.sanitized, sanitized)
+			if test.err {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
```
Comment (lenny-goodell) [inline internal/driver/macmapper_test.go]:
```suggestion
			if test.err {
				require.Error(t, err)
                                return
			require.NoError(t, err)
			assert.Equal(t, test.sanitized, sanitized)
```

**Review by lenny-goodell — DISMISSED** [review]

File: internal/driver/credentials.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+			err.Error(),
+			timer.RemainingAsString())
+		timer.SleepForInterval()
+	}
+
+	return credentials, err
+}
+
+func (d *Driver) getCredentialsFromMac(mac string) (Credentials, errors.EdgeX) {
+	if mac == "" {
```
Comment (ajcasagrande) [inline internal/driver/credentials.go]: This code needs to be a bit refactored. The defaults should be returned if the mac is empty _OR_ if the mac does not exist in the lookup table.

While refactoring, we may want to inline the `TryGetCredentialsForMACAddress` and make it part of this.

File: internal/driver/driver.go (hunk 4 above)
Comment (ajcasagrande) [reply internal/driver/driver.go:619]: Mostly because it was quick and dirty (which isn't a good excuse :wink: ), but also because that whole client and temporaryOnvifClient concept needs a good refactoring.

File: internal/driver/driver.go [hunk 7]
```diff
@@ -712,3 +690,45 @@ func (d *Driver) newTemporaryOnvifClient(device models.Device) (*OnvifClient, er
 	}
 	return client, nil
 }
+
+// refreshNetworkInterfaces will attempt to retrieve the device information for the specified camera
+// and update the values in the protocol properties
```
Comment (ajcasagrande) [inline internal/driver/driver.go]:
```suggestion
// refreshNetworkInterfaces will attempt to retrieve the mac address for the specified camera
// and update its value in the protocol properties
```

File: internal/driver/driver.go [hunk 8]
```diff
@@ -712,3 +690,45 @@ func (d *Driver) newTemporaryOnvifClient(device models.Device) (*OnvifClient, er
...
+
+	// update device to latest version in cache to prevent race conditions
+	device, edgeXErr := sdk.RunningService().GetDeviceByName(device.Name)
+	if err != nil {
+		return edgeXErr
+	}
+
+	device.Protocols[OnvifProtocol][MACAddress] = string(netInfo.NetworkInterfaces.Info.HwAddress)
+
+	return sdk.RunningService().UpdateDevice(device)
```
Comment (ajcasagrande) [inline internal/driver/driver.go]: we should check if it changed and only update the device if it has.

File: internal/driver/driver.go [hunk 9]
```diff
@@ -712,3 +690,45 @@ func (d *Driver) newTemporaryOnvifClient(device models.Device) (*OnvifClient, er
...
+	device, edgeXErr := sdk.RunningService().GetDeviceByName(device.Name)
+	if err != nil {
+		return edgeXErr
+	}
+
+	device.Protocols[OnvifProtocol][Manufacturer] = devInfo.Manufacturer
+	device.Protocols[OnvifProtocol][Model] = devInfo.Model
+	device.Protocols[OnvifProtocol][FirmwareVersion] = devInfo.FirmwareVersion
+	device.Protocols[OnvifProtocol][SerialNumber] = devInfo.SerialNumber
+	device.Protocols[OnvifProtocol][HardwareId] = devInfo.HardwareId
```
Comment (ajcasagrande) [inline internal/driver/driver.go]: same here, we should probably check if anything changed before sending a device update request.

File: internal/driver/macmapper.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+// inverts it into a quick lookup map of mac address to secret path.
+func (m *MACAddressMapper) UpdateMappings(raw map[string]string) {
+	m.credsMu.Lock()
+	defer m.credsMu.Unlock()
+
+	lc := sdk.RunningService().LoggingClient
+
+	credsMap := make(map[string]string)
+	for secretPath, macs := range raw {
+		if _, err := tryGetCredentials(secretPath); err != nil {
```
Comment (ajcasagrande) [inline internal/driver/macmapper.go]: This should probably just use the first part of `tryGetCredentials` which queries the `SecretProvider` directly. In the future it could make use of https://github.com/edgexfoundry/go-mod-bootstrap/issues/350 if implemented.

```suggestion
		if _, err := sdk.RunningService().SecretProvider.GetSecret(secretPath, UsernameKey, PasswordKey, AuthModeKey); err != nil {
```

File: internal/driver/macmapper.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+			lc.Warnf("One or more MAC address mappings exist for the secret path '%s' which does not exist in the Secret Store!", secretPath)
+		}
+
+		for _, mac := range strings.Split(macs, ",") {
+			sanitized, err := SanitizeMACAddress(mac)
+			if err != nil {
+				lc.Warnf("Skipping entry: %s", err.Error())
```
Comment (ajcasagrande) [inline internal/driver/macmapper.go]:
```suggestion
				lc.Warnf("Skipping invalid mac address %s: %s", mac, err.Error())
```

File: internal/driver/onvifdiscovery.go [hunk 12]
```diff
@@ -133,6 +132,13 @@ func (d *Driver) createDiscoveredDevice(onvifDevice onvif.Device) (sdkModel.Disc
 		device.Protocols[OnvifProtocol][DeviceStatus] = UpWithAuth
 		device.Protocols[OnvifProtocol][LastSeen] = time.Now().Format(time.UnixDate)
 
+		netInfo, err := d.getNetworkInterfaces(device)
+		if err != nil {
+			d.lc.Warnf("failed to get the network information for the camera %s, %v", endpointRefAddr, edgexErr)
```
Comment (ajcasagrande) [inline internal/driver/onvifdiscovery.go]: This should be moved below the deviceName name assignment, and the deviceName used in the error message instead of endpointRefAddr

File: internal/driver/macmapper.go [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+	i := 0
+	for mac := range m.credsMap {
+		macs[i] = mac
+		i++
+	}
+
+	return macs
+}
+
+func (m *MACAddressMapper) GetSecretPathForMACAddress(mac string) (string, error) {
```
Comment (ajcasagrande) [inline internal/driver/macmapper.go]: we should comment all exported functions

File: internal/driver/credentials.go [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+			secretPath,
+			err.Error(),
+			timer.RemainingAsString())
+		timer.SleepForInterval()
+	}
+
+	return credentials, err
+}
+
+func (d *Driver) getCredentialsFromMac(mac string) (Credentials, errors.EdgeX) {
```
Comment (ajcasagrande) [inline internal/driver/credentials.go]: this should ideally be separated from the driver code in such a way that it can be unit tested. Not sure how much work that is though.

File: internal/driver/credentials.go [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+	credentials := Credentials{
+		Username: secretData[UsernameKey],
+		Password: secretData[PasswordKey],
+		AuthMode: secretData[AuthModeKey],
+	}
+
+	if !IsAuthModeValid(secretData[AuthModeKey]) {
+		sdk.RunningService().LoggingClient.Warnf("AuthMode is set to an invalid value: %s. setting value to 'none'.", credentials.AuthMode)
+		credentials.AuthMode = AuthModeNone
```
Comment (ajcasagrande) [inline internal/driver/credentials.go]: Maybe we should default it to `usernametoken` or something sensible? If the creds are valid, but the mode isn't, I think we should still try and use them.

**Review by ajcasagrande — CHANGES_REQUESTED** [review]

File: internal/driver/macmapper.go [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+	if err != nil {
+		return Credentials{}, err
+	}
+	return tryGetCredentials(secretPath)
+}
+
+// SanitizeMACAddress takes in a MAC address in one of the IEEE 802 MAC-48, EUI-48, EUI-64 formats
+// and will return it in the standard go format, using colons and lower case letters:
+// Example:	aa:bb:cc:dd:ee:ff
+func SanitizeMACAddress(mac string) (string, error) {
```
Comment (ChristianDarr-personal) [inline internal/driver/macmapper.go:107]: I would add white-space trimming to this, because if the macs are listed comma separated with spaces `("aa:bb:cc:dd:ee:ff, 11:22:33:44:55:66")`, this will throw an error I believe.

**Review by ChristianDarr-personal — COMMENTED** [review]
Looks good, I just had one small comment.

File: internal/driver/driver.go [hunk 17]
```diff
@@ -712,3 +690,59 @@ func (d *Driver) newTemporaryOnvifClient(device models.Device) (*OnvifClient, er
...
+	if err != nil {
+		return edgeXErr
+	}
+
+	if devInfo.Manufacturer != device.Protocols[OnvifProtocol][Manufacturer] ||
+		devInfo.Model != device.Protocols[OnvifProtocol][Model] ||
+		devInfo.FirmwareVersion != device.Protocols[OnvifProtocol][FirmwareVersion] ||
+		devInfo.SerialNumber != device.Protocols[OnvifProtocol][SerialNumber] ||
+		devInfo.HardwareId != device.Protocols[OnvifProtocol][HardwareId] {
+		device.Protocols[OnvifProtocol][Manufacturer] = devInfo.Manufacturer
```
Comment (ajcasagrande) [inline internal/driver/driver.go:737]: the way go prefers to indent these makes it a bit confusing imo. I think a blank line after the `{` would make it more readable. and you can also get rid of the blank line before the updateDevice call.

File: internal/driver/macmapper_test.go [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+			err: true,
+		},
+		{
+			mac:       "AA:BB:CC:DD:EE:FF",
+			sanitized: "aa:bb:cc:dd:ee:ff",
+		},
+		{
+			mac:       "11-22-33-44-55-66",
+			sanitized: "11:22:33:44:55:66",
+		},
```
Comment (ajcasagrande) [inline internal/driver/macmapper_test.go:45]: add some test cases for the whitespace trimming

File: internal/driver/onvifclient.go [hunk 19]
```diff
@@ -60,12 +60,9 @@ func (d *Driver) newOnvifClient(device models.Device) (*OnvifClient, errors.Edge
 		return nil, errors.NewCommonEdgeX(errors.KindServerError, fmt.Sprintf("failed to create cameraInfo for camera %s", device.Name), edgexErr)
 	}
 
-	var credential config.Credentials
-	if cameraInfo.AuthMode != onvif.NoAuth {
-		credential, edgexErr = d.getCredentials(cameraInfo.SecretPath)
-		if edgexErr != nil {
-			return nil, errors.NewCommonEdgeX(errors.KindServerError, fmt.Sprintf("failed to get credentials for camera %s", device.Name), edgexErr)
-		}
+	credential, edgexErr := d.tryGetCredentialsFromMac(device.Protocols[OnvifProtocol][MACAddress])
```
Comment (ajcasagrande) [inline internal/driver/onvifclient.go]: I think this should be removed for now and put in a new branch for next sprint.

File: internal/driver/onvifdiscovery.go [hunk 20]
```diff
@@ -133,12 +132,20 @@ func (d *Driver) createDiscoveredDevice(onvifDevice onvif.Device) (sdkModel.Disc
 		device.Protocols[OnvifProtocol][DeviceStatus] = UpWithAuth
 		device.Protocols[OnvifProtocol][LastSeen] = time.Now().Format(time.UnixDate)
 
+		netInfo, err := d.getNetworkInterfaces(device)
```
Comment (ajcasagrande) [inline internal/driver/onvifdiscovery.go]: this should be moved down lower by where the err is actually checked.

File: internal/driver/checkstatuses.go (hunk 1 above)
Comment (ajcasagrande) [inline internal/driver/checkstatuses.go]: im not sure what this todo is for. maybe it needs to be removed?

File: internal/driver/credentials.go [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+			timer.RemainingAsString())
+		timer.SleepForInterval()
+	}
+
+	return credentials, err
+}
+
+// tryGetCredentialsFromMac will attempt to retrieve the credentials associated with the given mac address.
+func (d *Driver) tryGetCredentialsFromMac(mac string) (Credentials, errors.EdgeX) {
+	if mac == "" {
```
Comment (ajcasagrande) [inline internal/driver/credentials.go]: this still isn't what I had in mind, as there is a lot of duplicated code. this is also why i meantioned inlining the `TryGetCredentialsForMACAddress` as maybe you can just use the different flows to define the `secretPath`, and then they all flow into one block of code that does the lookup.

**Review by ajcasagrande — CHANGES_REQUESTED** [review]

File: internal/driver/checkstatuses.go (hunk 1 above)
Comment (ajcasagrande) [reply internal/driver/checkstatuses.go]: Yeah I'm not so sure what that comment referred to. Each of the sub calls below, and the one above update core-metadata, so i don't think it is that.

I think it might have been a todo written before the code that follows it (ie. should we refresh the device information and network information upon a device coming back online), which has been done.

File: internal/driver/credentials.go (hunk 21 above)
Comment (ajcasagrande) [reply internal/driver/credentials.go]: also, the mac needs to be sanitized first

**Review by ChristianDarr-personal — COMMENTED** [review]
LGTM

File: internal/driver/checkstatuses.go [hunk 22]
```diff
@@ -85,7 +100,7 @@ func (d *Driver) tcpProbe(device sdkModel.Device) bool {
 	return true
 }
 
-func (d *Driver) updateDeviceStatus(deviceName string, status string) error {
+func (d *Driver) updateDeviceStatus(deviceName string, status string) (bool, error) {
```
Comment (ajcasagrande) [inline internal/driver/checkstatuses.go:113]: add a function comment to say that it returns whether or not the device status changed, and any errors that ocurred.

File: internal/driver/driver.go (hunk 4 above)
Comment (ajcasagrande) [reply internal/driver/driver.go:619]: Will address this todo as part of a future story.

File: internal/driver/macmapper.go [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+				credsMap[sanitized] = secretPath
+			}
+		}
+	}
+
+	m.credsMap = credsMap
+}
+
+// ListMACAddresses will return a slice of mac addresses that have been assigned credentials
+func (m *MACAddressMapper) ListMACAddresses() []string {
```
Comment (ajcasagrande) [inline internal/driver/macmapper.go:65]: Note that this was part of checking off the acceptence criteria, but the function isn't called, so not sure if we even need it.

File: internal/driver/checkstatuses.go [hunk 24]
```diff
@@ -25,8 +25,23 @@ func (d *Driver) checkStatuses() {
 
 		status := d.testConnectionMethods(device)
 
-		if err := d.updateDeviceStatus(device.Name, status); err != nil {
+		if statusChanged, err := d.updateDeviceStatus(device.Name, status); err != nil {
 			d.lc.Warnf("Could not update device status for device %s: %s", device.Name, err.Error())
+		} else if statusChanged && status == UpWithAuth {
+			d.lc.Infof("Device %s is now %s, refreshing the device information.", device.Name, UpWithAuth)
+			go func(device models.Device) {
+				err2 := d.refreshDeviceInformation(device)
```
Comment (ChristianDarr-personal) [inline internal/driver/checkstatuses.go]: Could this keep the name err? Or should the name be more descriptive if it does need to be named differently?

File: internal/driver/macmapper.go (hunk 23 above)
Comment (brianointel) [reply internal/driver/macmapper.go:65]: Todd clarified it will be needed for the next story 7236

**Review by ajcasagrande — DISMISSED** [review]
LGTM

**Review by ajcasagrande — APPROVED** [review]
LGTM
